### PR TITLE
read druid query response by page

### PR DESCRIPTION
**Pull request by mqiang**
```
read druid query response by page
to solve issures: https://github.com/prestodb/presto/issues/14440
```

### CODE REVIEW

File: presto-druid/src/main/java/com/facebook/presto/druid/DruidClient.java [hunk 1]
```diff
@@ -142,7 +150,41 @@ private Request prepareQuery(String query)
...
                 .build();
     }
+
+    private static class StreamingJsonResponseHandler
+            implements ResponseHandler<InputStream, RuntimeException>
+    {
+        @Override
+        public InputStream handleException(Request request, Exception exception)
+        {
+            throw new RuntimeException("Request to worker failed", exception);
```
Comment (zhenxiao) [inline presto-druid/src/main/java/com/facebook/presto/druid/DruidClient.java:173]: shall we use PrestoException for StreamingJsonResponseHandler?

File: presto-druid/src/main/java/com/facebook/presto/druid/DruidClient.java [hunk 2]
```diff
@@ -119,21 +124,26 @@ public DruidSegmentInfo getSingleSegmentInfo(String dataSource, String segmentId
         return httpClient.execute(request, createJsonResponseHandler(SEGMENT_INFO_CODEC));
     }
 
-    public String getData(String dql)
+    public InputStream getData(String dql)
     {
-        return httpClient.execute(prepareQuery(dql), createStringResponseHandler()).getBody();
```
Comment (zhenxiao) [inline presto-druid/src/main/java/com/facebook/presto/druid/DruidClient.java:124]: I would suggest, rename prepareQuery to prepareMetadataQuery
and rename prepareQueryWithLines to prepareDataQuery
what do you think?

File: presto-druid/src/main/java/com/facebook/presto/druid/DruidClient.java [hunk 3]
```diff
@@ -142,7 +152,41 @@ private Request prepareQuery(String query)
 
         return setContentTypeHeaders(preparePost())
                 .setUri(uriBuilder.build())
-                .setBodyGenerator(createStaticBodyGenerator(createRequestBody(query)))
+                .setBodyGenerator(createStaticBodyGenerator(createRequestBody(query, "object", true)))
```
Comment (zhenxiao) [inline presto-druid/src/main/java/com/facebook/presto/druid/DruidClient.java:155]: "object" and "arrayLines" as constant? Or could we import from Druid code?

File: presto-druid/src/main/java/com/facebook/presto/druid/DruidBrokerPageSource.java [hunk 4]
```diff
@@ -151,21 +149,36 @@ else if (type instanceof TimestampType) {
...
+            }
+
             Page page = pageBuilder.build();
             completedPositions += page.getPositionCount();
-            finished = true;
+            completedBytes += page.getSizeInBytes();
+            pageBuilder.reset();
             return page;
         }
+        catch (Exception e) {
```
Comment (zhenxiao) [inline presto-druid/src/main/java/com/facebook/presto/druid/DruidBrokerPageSource.java:178]: is Exception too broad? I would suggest IOException, what do you think?

**Comment by zhenxiao** [comment]
btw, not sure, could we get rid of the first empty commit:
Merge pull request #1 from prestodb/master

**Comment by mqiang** [comment]
> btw, not sure, could we get rid of the first empty commit:
> Merge pull request #1 from prestodb/master

I don't know how to remove the first empty commit, or i can submit a new pull request

**Pull request closed without merging**